ElectronicObserver/Data/Battle/BattleRankPrediction.cs

```diff
@@ -0,0 +1,287 @@
+			if (hpAfter is null) continue;
+
+			EnemyHpBefore += hpBefore;
+			EnemyHpAfter += Math.Max((int)hpAfter, 0);
+			EnemyShipCount++;
+
+			if (hpAfter <= 0)
+			{
+				EnemyShipSunk++;
+			}
+		}
+	}
+
+	private void CalculateEnemyEscortFleetHp()
+	{
+		if (EnemyEscortFleetBefore is null) return;
+		if (EnemyEscortFleetAfter is null) return;
+
+		for (int index = 0; index < EnemyEscortFleetBefore.MembersInstance.Count; index++)
+		{
+			IShipData? ship = EnemyEscortFleetBefore.MembersInstance[index];
+
+			if (ship is null) continue;
+			if (!ship.CanBeTargeted) continue;
+
+			int hpBefore = ship.HPCurrent;
+			if (hpBefore < 0) continue;
+
+			int? hpAfter = EnemyEscortFleetAfter.MembersInstance[index]?.HPCurrent;
+			if (hpAfter is null) continue;
+
+			EnemyHpBefore += hpBefore;
+			EnemyHpAfter += Math.Max((int)hpAfter, 0);
+			EnemyShipCount++;
+
+			if (hpAfter <= 0)
+			{
+				EnemyShipSunk++;
+			}
+		}
+	}
+
+	private void ResetValues()
+	{
+		FriendlyShipCount = 0;
+		FriendlyShipSunk = 0;
+		FriendlyHpBefore = 0;
+		FriendlyHpAfter = 0;
+
+		EnemyShipCount = 0;
+		EnemyShipSunk = 0;
+		EnemyHpBefore = 0;
+		EnemyHpAfter = 0;
+	}
+
+	/// <summary>
+	/// Create a copy of the fleet with their HP after the battle
+	/// </summary>
+	/// <param name="fleet"></param>
+	/// <param name="resultHp"></param>
+	/// <param name="battleSide"></param>
+	/// <returns></returns>
+	public static IFleetData? SimulateFleetAfterBattle(IFleetData? fleet, List<int> resultHp, BattleSides battleSide)
+	{
+		if (fleet is null) return null;
+
+		int offset = BattleIndex.Get(battleSide, 0);
+		List<int> resultHpWithOffset = resultHp.Skip(offset).Take(fleet.MembersInstance.Count).ToList();
+
+		return SimulateFleetAfterBattle(fleet, resultHpWithOffset);
+	}
+
+	/// <summary>
+	/// Create a copy of the fleet with their HP after the battle
+	/// </summary>
+	/// <param name="fleet"></param>
+	/// <param name="resultHp"></param>
+	/// <returns></returns>
+	public static IFleetData SimulateFleetAfterBattle(IFleetData fleet, List<int> resultHp)
+	{
+		IFleetData fleetClone = fleet.DeepClone();
+
+		if (fleetClone.Members is null) return fleetClone;
+		if (resultHp.Count < fleetClone.Members.Count) return fleetClone;
+
+		for (int index = 0; index < fleetClone.Members.Count; index++)
+		{
+			if (fleetClone.MembersInstance[index] is {} ship)
+			{
+				ship.HPCurrent = resultHp[index];
+			}
+		}
+
+		return fleetClone;
+	}
+	
+	private BattleRank GetWinRank()
+	{
+		int rifriend = (int)(FriendHpRate * 100);
+		int rienemy = (int)(EnemyHpRate * 100);
+
+		// 轟沈艦なし
+		if (FriendlyShipSunk == 0)
+		{
+			// 敵艦全撃沈
+			if (EnemyShipSunk == EnemyShipCount)
+			{
+				return FriendHpRate switch
+				{
+					<= 0 => BattleRank.SS,
+					_ => BattleRank.S,
+				};
+			}
+			
+			if (EnemyShipCount > 1 && EnemyShipSunk >= (int)(EnemyShipCount * 0.7)) // 敵の 70% 以上を撃沈
+			{
+				return BattleRank.A;   // A
```

Comment:
You can remove the A comment here, the enum explains it now.

Reply:
Same could be said for all comments in this function, removing them

Reply:
Yeah.